## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#44417eac-ec93-4fdc-aed8-c2ec40e0c88f) |
| mint-compression-dist-xl.sh | :x: [more...](#80b07f81-4a76-4647-8a8d-82ed4b10556b) |
<a name="80b07f81-4a76-4647-8a8d-82ed4b10556b"></a>
### 8475-196c35f/mint-compression-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.59:30490
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp b76ee3509378:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 2 seconds
(3/14) Running aws-sdk-php tests ... done in 45 seconds
(4/14) Running aws-sdk-ruby tests ... done in 3 seconds
(5/14) Running awscli tests ... done in 2 minutes and 9 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 54 seconds
(8/14) Running minio-dotnet tests ... done in 39 seconds
(9/14) Running minio-go tests ... done in 56 seconds
(10/14) Running minio-java tests ... done in 1 minutes and 37 seconds
(11/14) Running minio-js tests ... done in 56 seconds
(12/14) Running minio-py tests ... FAILED in 2 minutes and 58 seconds
{
  "message": "'content-language'",
  "function": "presigned_get_object(bucket_name, object_name, expires, response_headers, request_date)",
  "args": {
    "object_name": "0d19d7d3-0f31-4a2f-bd3d-f42f139232d3",
    "presigned_get_object_url": "http://72.28.97.59:30490/minio-py-test-6abb95c8-8063-40b5-9672-20a42a583fbc/0d19d7d3-0f31-4a2f-bd3d-f42f139232d3?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=minio%2F20191101%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20191101T203633Z&X-Amz-Expires=5&X-Amz-SignedHeaders=host&response-content-language=en_US&response-content-type=text%2Fplain&X-Amz-Signature=3d8323ede1fedacea4b09de65e006d4b374e4f94d898bbf51fa1b99f4b86f923",
    "bucket_name": "minio-py-test-6abb95c8-8063-40b5-9672-20a42a583fbc"
  },
  "name": "minio-py:test_presigned_get_object_response_headers",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1396, in test_presigned_get_object_response_headers\n    returned_content_language = response.headers['Content-Language']\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/_collections.py\", line 156, in __getitem__\n    val = self._container[key.lower()]\nKeyError: 'content-language'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 2039, in main\n    test_presigned_get_object_response_headers(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1414, in test_presigned_get_object_response_headers\n    raise Exception(err)\nException: 'content-language'\n",
  "status": "FAIL",
  "duration": 101
}

Executed 11 out of 14 tests successfully.
```
<a name="44417eac-ec93-4fdc-aed8-c2ec40e0c88f"></a>
### 8475-196c35f/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.59:30475
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp cc1af6ba3573:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 2 seconds
(2/14) Running aws-sdk-java tests ... done in 2 seconds
(3/14) Running aws-sdk-php tests ... done in 1 minutes and 3 seconds
(4/14) Running aws-sdk-ruby tests ... done in 12 seconds
(5/14) Running awscli tests ... done in 2 minutes and 39 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 1 minutes and 0 seconds
(8/14) Running minio-dotnet tests ... done in 1 minutes and 27 seconds
(9/14) Running minio-go tests ... FAILED in 4 minutes and 5 seconds
{
  "args": {},
  "duration": 89929,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 8 out of 14 tests successfully.
```
Error: No such image: minio/minio:8475-196c35f

--- a/cmd/gateway-common.go
+++ b/cmd/gateway-common.go
@@ -17,8 +17,10 @@
 package cmd
 
 import (
+	"context"
 	"net/http"
 	"strings"
+	"time"
 
 	"github.com/minio/minio/cmd/config"
 	xhttp "github.com/minio/minio/cmd/http"
@@ -285,6 +287,28 @@ func ToMinioClientCompleteParts(parts []CompletePart) []minio.CompletePart {
 		mparts[i] = ToMinioClientCompletePart(part)
 	}
 	return mparts
+}
+
+// IsBackendOnline - verifies if the backend is reachable
+// by performing a GET request on the URL. returns 'true'
+// if backend is reachable.
+func IsBackendOnline(ctx context.Context, clnt *http.Client, urlStr string) bool {
+	ctx, cancel := context.WithTimeout(ctx, 1*time.Second)
+	defer cancel()
+
+	// never follow redirects
+	clnt.CheckRedirect = func(*http.Request, []*http.Request) error {
+		return http.ErrUseLastResponse
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, urlStr, nil)
+	if err != nil {
+		return false
+	}
+	if _, err = clnt.Do(req); err != nil {
+		return !xnet.IsNetworkOrHostDown(err)
+	}
+	return true
 }
 
 // ErrorRespToObjectError converts MinIO errors to minio object layer errors.

--- a/cmd/gateway/azure/gateway-azure.go
+++ b/cmd/gateway/azure/gateway-azure.go
@@ -185,7 +185,9 @@ func (g *Azure) NewGatewayLayer(creds auth.Credentials) (minio.ObjectLayer, erro
 	c.HTTPClient = &http.Client{Transport: minio.NewCustomHTTPTransport()}
 
 	return &azureObjects{
-		client: c.GetBlobService(),
+		endpoint:   fmt.Sprintf("https://%s.blob.core.windows.net", creds.AccessKey),
+		httpClient: c.HTTPClient,
+		client:     c.GetBlobService(),
 	}, nil
 }
 
@@ -343,7 +345,9 @@ func azurePropertiesToS3Meta(meta storage.BlobMetadata, props storage.BlobProper
 // azureObjects - Implements Object layer for Azure blob storage.
 type azureObjects struct {
 	minio.GatewayUnsupported
-	client storage.BlobStorageClient // Azure sdk client
+	endpoint   string
+	httpClient *http.Client
+	client     storage.BlobStorageClient // Azure sdk client
 }
 
 // Convert azure errors to minio object layer errors.
@@ -447,6 +451,8 @@ func (a *azureObjects) Shutdown(ctx context.Context) error {
 
 // StorageInfo - Not relevant to Azure backend.
 func (a *azureObjects) StorageInfo(ctx context.Context) (si minio.StorageInfo) {
+	si.Backend.Type = minio.BackendGateway
+	si.Backend.GatewayOnline = minio.IsBackendOnline(ctx, a.httpClient, a.endpoint)
 	return si
 }
 

--- a/cmd/gateway/b2/gateway-b2.go
+++ b/cmd/gateway/b2/gateway-b2.go
@@ -123,7 +123,10 @@ func (g *B2) NewGatewayLayer(creds auth.Credentials) (minio.ObjectLayer, error) 
 	return &b2Objects{
 		creds:    creds,
 		b2Client: client,
-		ctx:      ctx,
+		httpClient: &http.Client{
+			Transport: minio.NewCustomHTTPTransport(),
+		},
+		ctx: ctx,
 	}, nil
 }
 
@@ -135,10 +138,11 @@ func (g *B2) Production() bool {
 // b2Object implements gateway for MinIO and BackBlaze B2 compatible object storage servers.
 type b2Objects struct {
 	minio.GatewayUnsupported
-	mu       sync.Mutex
-	creds    auth.Credentials
-	b2Client *b2.B2
-	ctx      context.Context
+	mu         sync.Mutex
+	creds      auth.Credentials
+	b2Client   *b2.B2
+	httpClient *http.Client
+	ctx        context.Context
 }
 
 // Convert B2 errors to minio object layer errors.
@@ -225,12 +229,13 @@ func b2MsgCodeToObjectError(code int, msgCode string, msg string, params ...stri
 // Shutdown saves any gateway metadata to disk
 // if necessary and reload upon next restart.
 func (l *b2Objects) Shutdown(ctx context.Context) error {
-	// TODO
 	return nil
 }
 
 // StorageInfo is not relevant to B2 backend.
 func (l *b2Objects) StorageInfo(ctx context.Context) (si minio.StorageInfo) {
+	si.Backend.Type = minio.BackendGateway
+	si.Backend.GatewayOnline = minio.IsBackendOnline(ctx, l.httpClient, "https://api.backblazeb2.com/b2api/v1")
 	return si
 }
 

--- a/cmd/gateway/gcs/gateway-gcs.go
+++ b/cmd/gateway/gcs/gateway-gcs.go
@@ -208,6 +208,9 @@ func (g *GCS) NewGatewayLayer(creds auth.Credentials) (minio.ObjectLayer, error)
 	gcs := &gcsGateway{
 		client:    client,
 		projectID: g.projectID,
+		httpClient: &http.Client{
+			Transport: minio.NewCustomHTTPTransport(),
+		},
 	}
 
 	// Start background process to cleanup old files in minio.sys.tmp
@@ -346,8 +349,9 @@ func isValidGCSProjectIDFormat(projectID string) bool {
 // gcsGateway - Implements gateway for MinIO and GCS compatible object storage servers.
 type gcsGateway struct {
 	minio.GatewayUnsupported
-	client    *storage.Client
-	projectID string
+	client     *storage.Client
+	httpClient *http.Client
+	projectID  string
 }
 
 // Returns projectID from the GOOGLE_APPLICATION_CREDENTIALS file.
@@ -412,8 +416,10 @@ func (l *gcsGateway) Shutdown(ctx context.Context) error {
 }
 
 // StorageInfo - Not relevant to GCS backend.
-func (l *gcsGateway) StorageInfo(ctx context.Context) minio.StorageInfo {
-	return minio.StorageInfo{}
+func (l *gcsGateway) StorageInfo(ctx context.Context) (si minio.StorageInfo) {
+	si.Backend.Type = minio.BackendGateway
+	si.Backend.GatewayOnline = minio.IsBackendOnline(ctx, l.httpClient, "https://storage.googleapis.com")
+	return si
 }
 
 // MakeBucketWithLocation - Create a new container on GCS backend.

--- a/cmd/gateway/hdfs/gateway-hdfs.go
+++ b/cmd/gateway/hdfs/gateway-hdfs.go
@@ -227,7 +227,8 @@ func (n *hdfsObjects) StorageInfo(ctx context.Context) minio.StorageInfo {
 	}
 	sinfo := minio.StorageInfo{}
 	sinfo.Used = []uint64{fsInfo.Used}
-	sinfo.Backend.Type = minio.Unknown
+	sinfo.Backend.Type = minio.BackendGateway
+	sinfo.Backend.GatewayOnline = true
 	return sinfo
 }
 

--- a/cmd/gateway/nas/gateway-nas.go
+++ b/cmd/gateway/nas/gateway-nas.go
@@ -125,7 +125,8 @@ func (n *nasObjects) IsListenBucketSupported() bool {
 
 func (n *nasObjects) StorageInfo(ctx context.Context) minio.StorageInfo {
 	sinfo := n.ObjectLayer.StorageInfo(ctx)
-	sinfo.Backend.Type = minio.Unknown
+	sinfo.Backend.GatewayOnline = sinfo.Backend.Type == minio.BackendFS
+	sinfo.Backend.Type = minio.BackendGateway
 	return sinfo
 }
 

--- a/cmd/gateway/oss/gateway-oss.go
+++ b/cmd/gateway/oss/gateway-oss.go
@@ -145,7 +145,9 @@ func (g *OSS) NewGatewayLayer(creds auth.Credentials) (minio.ObjectLayer, error)
 	if err != nil {
 		return nil, err
 	}
-
+	client.HTTPClient = &http.Client{
+		Transport: minio.NewCustomHTTPTransport(),
+	}
 	return &ossObjects{
 		Client: client,
 	}, nil
@@ -341,7 +343,9 @@ func (l *ossObjects) Shutdown(ctx context.Context) error {
 
 // StorageInfo is not relevant to OSS backend.
 func (l *ossObjects) StorageInfo(ctx context.Context) (si minio.StorageInfo) {
-	return
+	si.Backend.Type = minio.BackendGateway
+	si.Backend.GatewayOnline = minio.IsBackendOnline(ctx, l.Client.HTTPClient, l.Client.Config.Endpoint)
+	return si
 }
 
 // ossIsValidBucketName verifies whether a bucket name is valid.

--- a/cmd/gateway/s3/gateway-s3.go
+++ b/cmd/gateway/s3/gateway-s3.go
@@ -261,7 +261,11 @@ func (g *S3) NewGatewayLayer(creds auth.Credentials) (minio.ObjectLayer, error) 
 
 	s := s3Objects{
 		Client: clnt,
+		HTTPClient: &http.Client{
+			Transport: minio.NewCustomHTTPTransport(),
+		},
 	}
+
 	// Enables single encryption of KMS is configured.
 	if minio.GlobalKMS != nil {
 		encS := s3EncObjects{s}
@@ -283,7 +287,8 @@ func (g *S3) Production() bool {
 // s3Objects implements gateway for MinIO and S3 compatible object storage servers.
 type s3Objects struct {
 	minio.GatewayUnsupported
-	Client *miniogo.Core
+	Client     *miniogo.Core
+	HTTPClient *http.Client
 }
 
 // Shutdown saves any gateway metadata to disk
@@ -294,6 +299,8 @@ func (l *s3Objects) Shutdown(ctx context.Context) error {
 
 // StorageInfo is not relevant to S3 backend.
 func (l *s3Objects) StorageInfo(ctx context.Context) (si minio.StorageInfo) {
+	si.Backend.Type = minio.BackendGateway
+	si.Backend.GatewayOnline = minio.IsBackendOnline(ctx, l.HTTPClient, l.Client.EndpointURL().String())
 	return si
 }
 

--- a/cmd/healthcheck-handler.go
+++ b/cmd/healthcheck-handler.go
@@ -59,18 +59,15 @@ func LivenessCheckHandler(w http.ResponseWriter, r *http.Request) {
 		// is able to start on orchestration platforms like Docker Swarm.
 		// Refer https://github.com/minio/minio/issues/8140 for more details.
 		// Make sure to add server not initialized status in header
-		w.Header().Set(xhttp.MinIOServerStatus, "Server-not-initialized")
+		w.Header().Set(xhttp.MinIOServerStatus, "server-not-initialized")
 		writeSuccessResponseHeadersOnly(w)
 		return
 	}
 
 	if !globalIsXL && !globalIsDistXL {
 		s := objLayer.StorageInfo(ctx)
-		// Gateways don't provide disk info.
-		if s.Backend.Type == Unknown {
-			// ListBuckets to confirm gateway backend is up
-			if _, err := objLayer.ListBuckets(ctx); err != nil {
-				logger.LogOnceIf(ctx, err, struct{}{})
+		if s.Backend.Type == BackendGateway {
+			if !s.Backend.GatewayOnline {
 				writeResponse(w, http.StatusServiceUnavailable, nil, mimeNone)
 				return
 			}

--- a/cmd/object-api-datatypes.go
+++ b/cmd/object-api-datatypes.go
@@ -34,6 +34,8 @@ const (
 	BackendFS
 	// Multi disk BackendErasure (single, distributed) backend.
 	BackendErasure
+	// Gateway backend.
+	BackendGateway
 	// Add your own backend.
 )
 
@@ -49,8 +51,11 @@ type StorageInfo struct {
 
 	// Backend type.
 	Backend struct {
-		// Represents various backend types, currently on FS and Erasure.
+		// Represents various backend types, currently on FS, Erasure and Gateway
 		Type BackendType
+
+		// Following fields are only meaningful if BackendType is Gateway.
+		GatewayOnline bool
 
 		// Following fields are only meaningful if BackendType is Erasure.
 		OnlineDisks      madmin.BackendDisks // Online disks during server startup.

